Add option to use HSI as clock source for stm32 devices

fmanno opened the pull request:
- Add new compile flag MICROPY_HW_CLK_SRC_HSI=1
  - Modify documentation to include HSI

fmanno committed:
Add option to use HSI as clock source
  * Add new compile flag MICROPY_HW_CLK_SRC_HSI=1
  * Modify documentation to include HSI

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -347,6 +347,10 @@ SRC_USBDEV = $(addprefix $(USBDEV_DIR)/,\
 
+ifeq ($(MICROPY_HW_CLK_SRC_HSI),1)
+CFLAGS_MOD += -DMICROPY_HW_CLK_SRC_HSI=1
+endif
```
It doesn't look like `MICROPY_HW_CLK_SRC_HSI` needs to be something you set via the Makefile (eg it doesn't include/exclude certain .c files in the build).  For options that don't need to be via the Makefile it's best to keep them just internal to C.  Basically these lines are not needed.

dpgeorge commented on ports/stm32/boards/NUCLEO_F401RE/mpconfigboard.h:
```diff
@@ -5,8 +5,15 @@
+#define MICROPY_HW_CLK_PLLM (16)
+#else
 #define MICROPY_HW_CLK_PLLM (8)
+#endif
```
I think it's fair enough to have the default for this board as using HSI, as an example of how it works.  Then a comment saying how to change to HSE if needed.

dpgeorge commented on ports/stm32/mpconfigboard_common.h:
```diff
@@ -181,11 +181,26 @@
-#else
-#define MICROPY_HW_CLK_HSE_STATE (RCC_HSE_ON)
+// Configure clock source for the chip
+#if MICROPY_HW_CLK_SRC_HSI
```
To be consistent with other config names, probably better to call this `MICROPY_HW_CLK_USE_HSI` instead (see eg `MICROPY_HW_RTC_USE_LSE`).

dpgeorge commented on ports/stm32/mpconfigboard_common.h:
```diff
@@ -181,11 +181,26 @@
-#define MICROPY_HW_CLK_HSE_STATE (RCC_HSE_ON)
+// Configure clock source for the chip
+#if MICROPY_HW_CLK_SRC_HSI
+#define MICROPY_RCC_OSCILLATOR_TYPE (RCC_OSCILLATORTYPE_HSI)
```
To keep the namespaces separated, please use `MICROPY_HW_RCC_xxx`.

dpgeorge commented on ports/stm32/system_stm32.c:
```diff
@@ -538,7 +540,7 @@ void SystemClock_Config(void)
 #endif
 
-    uint32_t vco_out = RCC_OscInitStruct.PLL.PLLN * (HSE_VALUE / 1000000) / RCC_OscInitStruct.PLL.PLLM;
+    uint32_t vco_out = RCC_OscInitStruct.PLL.PLLN * (HSI_VALUE / 1000000) / RCC_OscInitStruct.PLL.PLLM;
```
This is going to need a config or something because `HSE_VALUE` should be used if HSE is used.

dpgeorge reviewed (commented):
Thanks for the contribution!  It's a good improvement.

fmanno committed:
Address review comments
- Remove compile flag from Makefile
- Make HSI the default for NUCLEO_F401RE board
- Add comments with instructions to use HSE
- Add missing define to hold either HSE or HSI value

fmanno commented:
I've hopefully addressed your comments. I've also updated the wiki for this board

dpgeorge commented:
Thanks for updating.  I made a few more minor edits (indenting and renamed existing MICROPY_HW_CLK_HSE_STATE to MICROPY_HW_RCC_HSE_STATE to make it consistent) and merged it in f938e70c6905474b08f7b5e6df3da9c1b0dcf01a
